src/backend/OpenCL/CHIPBackendOpenCL.cc

```diff
@@ -1260,46 +1320,430 @@ CHIPQueueOpenCL::enqueueBarrierImpl(std::vector<CHIPEvent *> *EventsToWaitFor) {
+                                                      nullptr, &TmpEv);
+  CHIPERR_CHECK_LOG_AND_THROW(Status, CL_SUCCESS, hipErrorTbd);
+  Event->reset(TmpEv);
+  return Event;
+}
+
+void CHIPQueueOpenCL::destroyNativeGraph(CHIPGraphNative *NativeGraph) {
+  if (NativeGraph == nullptr)
+    return;
+  CHIPGraphNativeOpenCL *G = (CHIPGraphNativeOpenCL *)NativeGraph;
+  delete G;
+}
+
+bool CHIPGraphNativeOpenCL::addNode(CHIPGraphNode *NewNode) {
+  cl_sync_point_khr NewSyncPoint = -1;
+
+  // map the dependent CHIPGraphNodes to OpenCL syncpoints
+  const std::vector<CHIPGraphNode *> &Dependencies = NewNode->getDependencies();
+  std::vector<cl_sync_point_khr> SyncPointDeps;
+  for (auto Node : Dependencies) {
+    auto Iter = SyncPointMap.find(Node);
+    if (Iter == SyncPointMap.end()) {
+      logError("Can't find SyncPoint for Node");
+      return false;
+    }
+    SyncPointDeps.push_back(Iter->second);
+  }
+
+  hipGraphNodeType NodeType = NewNode->getType();
+  bool Res;
+  switch (NodeType) {
+  case hipGraphNodeTypeKernel:
+    Res = addKernelNode((CHIPGraphNodeKernel *)NewNode, SyncPointDeps,
+                        &NewSyncPoint);
+    break;
+  case hipGraphNodeTypeEmpty:
+    assert(0 && "Empty node should be removed earlier");
+
+#ifdef cl_pocl_command_buffer_svm
+  case hipGraphNodeTypeMemcpy:
+    Res = addMemcpyNode((CHIPGraphNodeMemcpy *)NewNode, SyncPointDeps,
+                        &NewSyncPoint);
+    break;
+  case hipGraphNodeTypeMemset:
+    Res = addMemsetNode((CHIPGraphNodeMemset *)NewNode, SyncPointDeps,
+                        &NewSyncPoint);
+    break;
+  case hipGraphNodeTypeMemcpyFromSymbol:
+    Res = addMemcpyNode((CHIPGraphNodeMemcpyFromSymbol *)NewNode, SyncPointDeps,
+                        &NewSyncPoint);
+    break;
+  case hipGraphNodeTypeMemcpyToSymbol:
+    Res = addMemcpyNode((CHIPGraphNodeMemcpyToSymbol *)NewNode, SyncPointDeps,
+                        &NewSyncPoint);
+    break;
+#endif
+
+#ifdef cl_pocl_command_buffer_host_exec
+  case hipGraphNodeTypeWaitEvent:
+    Res = addEventWaitNode((CHIPGraphNodeWaitEvent *)NewNode, SyncPointDeps,
+                           &NewSyncPoint);
+    break;
+  case hipGraphNodeTypeEventRecord:
+    Res = addEventRecordNode((CHIPGraphNodeEventRecord *)NewNode, SyncPointDeps,
+                             &NewSyncPoint);
+    break;
+  case hipGraphNodeTypeHost:
+    Res =
+        addHostNode((CHIPGraphNodeHost *)NewNode, SyncPointDeps, &NewSyncPoint);
+    break;
+#endif
+
+  default:
+    Res = false;
+  }
+  if (!Res)
+    return false;
+
+  SyncPointMap.insert(std::make_pair(NewNode, NewSyncPoint));
+  return true;
+}
+
+bool CHIPGraphNativeOpenCL::finalize() {
+  int Status = Exts->clFinalizeCommandBufferKHR(Handle);
+  if (Status == CL_SUCCESS) {
+    Finalized = true;
+    return true;
+  } else {
+    logError("clFinalizeCommandBufferKHR FAILED with status {}",
+             resultToString(Status));
+    return false;
+  }
```

Comment:
Style nit (no else after return):
```suggestion
  if (Status == CL_SUCCESS) {
    Finalized = true;
    return true;
  }
  logError("clFinalizeCommandBufferKHR FAILED with status {}",
           resultToString(Status));
  return false;
```